File: dotnet/Pruvz.EvidencePacket/EvidenceLog.cs
// Append-only evidence log, format version 1 — independent .NET implementation
// (PRUVZ-97), from docs/EVIDENCE-LOG.md: RFC 6962 / RFC 9162 tree composition
// over domain-separated leaves, signed checkpoints, and the stateful
// acceptance rules.

using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace Pruvz.EvidencePacket;

public sealed class EvidenceLogException(string code, string message) : Exception(message)
{
    public string Code { get; } = code;
}

public static class EvidenceLog
{
    public const string FormatVersion = "1";
    public const string LeafEnvelopeVersion = "1";
    private const string LeafDomainTag = "pruvz.ai/evidence-log-leaf";
    private const string CheckpointDomainTag = "pruvz.ai/log-checkpoint";
    private const char Separator = '\u0000';
    private const long MaxSafeInteger = 9007199254740991;
    private const int MaxBoundTextLength = 512;

    public static readonly string[] AssuranceProfiles = ["PRE_CUSTOMER_DEFAULT", "CUSTOMER_PRODUCTION"];
    private static readonly Regex HexPattern = new("^[0-9a-f]{64}$", RegexOptions.Compiled);
    private static readonly Regex RootHashPattern = new("^sha256:[0-9a-f]{64}$", RegexOptions.Compiled);
    private static readonly Regex PrintableAscii = new("^[\\x20-\\x7e]+$", RegexOptions.Compiled);

    private static EvidenceLogException Refuse(string code, string message) => new(code, message);

    private static byte[] CanonicalBytes(JsonNode? document, string context)
    {
        try
        {
            return Canonical.Canonicalize(document);
        }
        catch (CommitmentException error)
        {
            throw Refuse("LOG_MALFORMED", $"{context} cannot be canonicalized: {error.Message}");
        }
    }

    private static byte[] Sha256(params byte[][] parts)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        foreach (var part in parts)
        {
            hash.AppendData(part);
        }
        return hash.GetHashAndReset();
    }

    private static byte[] HashFromHex(JsonNode? node, string field) => HashFromHex(Registry.AsString(node), field);

    private static byte[] HashFromHex(string? value, string field)
    {
        if (value is null || !HexPattern.IsMatch(value))
        {
            throw Refuse("LOG_MALFORMED", $"{field} must be 64 lowercase hex digits of SHA-256");
        }
        return Convert.FromHexString(value);
    }

    private static byte[] DecodeSignature(JsonNode? node, string field)
    {
        try
        {
            return Registry.DecodeBase64Url(node, field, "MALFORMED_SIGNATURE");
        }
        catch (TrustRegistryException error)
        {
            throw Refuse("MALFORMED_SIGNATURE", error.Message);
        }
    }

    private static string BoundText(JsonNode? node, string field)
    {
        var value = Registry.AsString(node);
        if (string.IsNullOrEmpty(value))
        {
            throw Refuse("LOG_MALFORMED", $"{field} is required");
        }
        if (value.Length > MaxBoundTextLength)
        {
            throw Refuse("LOG_MALFORMED", $"{field} must not exceed {MaxBoundTextLength} characters");
        }
        if (!PrintableAscii.IsMatch(value))
        {
            throw Refuse("LOG_MALFORMED", $"{field} must be printable ASCII");
        }
        return value;
    }

    private static long SafePositiveInteger(JsonNode? node, string field)
    {
        var value = Registry.IntegerOf(node);
        var isNumber = node is JsonValue leaf && leaf.TryGetValue<JsonElement>(out var element) && element.ValueKind == JsonValueKind.Number
            || node is JsonValue direct && direct.TryGetValue<long>(out _);
        if (!isNumber || value is null || value < 1 || value > MaxSafeInteger)
        {
            throw Refuse("LOG_MALFORMED", $"{field} must be an integer between 1 and {MaxSafeInteger}");
        }
        return value.Value;
    }

    public static void RequireSupported(string? formatVersion)
    {
        if (formatVersion != FormatVersion)
        {
            throw Refuse("UNKNOWN_LOG_VERSION", $"Unknown evidence-log version \"{formatVersion}\".");
        }
    }

    // -- Leaf encoding -------------------------------------------------------

    public static byte[] LeafInput(JsonNode? node)
    {
        if (node is not JsonObject seal || seal["envelope"] is not JsonObject envelope)
        {
            throw Refuse("LOG_MALFORMED", "a leaf must be { envelope, signature }");
        }
        if (Canonical.StringOf(envelope, "version") != LeafEnvelopeVersion)
        {
            throw Refuse("LOG_MALFORMED", $"a leaf carries an envelope of version {LeafEnvelopeVersion}");
        }
        var signature = DecodeSignature(seal["signature"], "seal.signature");
        var suiteId = envelope["signer"] is JsonObject signer ? Canonical.StringOf(signer, "suite") : null;
        if (suiteId is not null && Registry.Suites.TryGetValue(suiteId, out var suite) && signature.Length != suite.SignatureLength)
        {
            throw Refuse("MALFORMED_SIGNATURE", $"seal.signature is {signature.Length} bytes");
        }
        var unknown = seal.Select(pair => pair.Key).Where(member => member != "envelope" && member != "signature").ToList();
        if (unknown.Count > 0)
        {
            throw Refuse("LOG_MALFORMED", $"a leaf holds exactly envelope and signature; got {string.Join(", ", unknown)}");
        }
        var header = string.Join(Separator, LeafDomainTag, FormatVersion, "");
        return [.. Encoding.UTF8.GetBytes(header), .. CanonicalBytes(seal, "a leaf")];
    }

    public static byte[] LeafHashOf(byte[] input) => Sha256([0x00], input);

    public static string SealLeafHash(JsonNode? seal) =>
        Convert.ToHexString(LeafHashOf(LeafInput(seal))).ToLowerInvariant();

    // -- Tree composition ----------------------------------------------------

    private static byte[] Node(byte[] left, byte[] right) => Sha256([0x01], left, right);

    private static int Split(int n)
    {
        var k = 1;
        while (k * 2 < n)
        {
            k *= 2;
        }
        return k;
    }

    private static byte[] Head(List<byte[]> hashes, int lo, int hi)
    {
        var n = hi - lo;
        if (n == 1)
        {
            return hashes[lo];
        }
        var k = Split(n);
        return Node(Head(hashes, lo, lo + k), Head(hashes, lo + k, hi));
    }

    public static string TreeHead(IEnumerable<string> leafHashes)
    {
        var hashes = leafHashes.Select((value, index) => HashFromHex(value, $"leafHashes[{index}]")).ToList();
        if (hashes.Count == 0)
        {
            return Convert.ToHexString(SHA256.HashData(Array.Empty<byte>())).ToLowerInvariant();
        }
        return Convert.ToHexString(Head(hashes, 0, hashes.Count)).ToLowerInvariant();
    }

    public static void VerifyInclusion(string? leafHash, long? leafIndex, long? treeSize, JsonArray? path, string? rootHash)
    {
        var leaf = HashFromHex(leafHash, "leafHash");
        var root = HashFromHex(rootHash, "rootHash");
        if (leafIndex is null || leafIndex < 0)
        {
            throw Refuse("LOG_MALFORMED", "leafIndex must be a non-negative integer");
        }
        if (treeSize is null || treeSize < 1)
        {
            throw Refuse("LOG_MALFORMED", "treeSize must be a positive integer");
        }
        if (path is null)
        {
            throw Refuse("LOG_MALFORMED", "path must be an array of node hashes");
        }
        var nodes = path.Select((value, index) => HashFromHex(value, $"path[{index}]")).ToList();

        if (leafIndex >= treeSize)
        {
            throw Refuse("INCLUSION_PROOF_INVALID", $"leaf index {leafIndex} does not exist in a tree of {treeSize} leaves");
        }
        var fn = leafIndex.Value;
        var sn = treeSize.Value - 1;
        var r = leaf;
        foreach (var p in nodes)
        {
            if (sn == 0)
            {
                throw Refuse("INCLUSION_PROOF_INVALID", "the path is longer than the tree is deep");
            }
            if (fn % 2 == 1 || fn == sn)
            {
                r = Node(p, r);
                if (fn % 2 == 0)
                {
                    while (fn != 0 && fn % 2 == 0)
                    {
                        fn /= 2;
                        sn /= 2;
                    }
                }
            }
            else
            {
                r = Node(r, p);
            }
            fn /= 2;
            sn /= 2;
        }
        if (sn != 0)
        {
            throw Refuse("INCLUSION_PROOF_INVALID", "the path is shorter than the tree is deep");
        }
        if (!r.AsSpan().SequenceEqual(root))
        {
            throw Refuse("INCLUSION_PROOF_INVALID", "the path does not lead this leaf to this root");
        }
    }

    public static void VerifyConsistency(long? fromSize, string? fromRootHash, long? toSize, string? toRootHash, JsonArray? proof)
    {
        var fromRoot = HashFromHex(fromRootHash, "fromRootHash");
        var toRoot = HashFromHex(toRootHash, "toRootHash");
        if (fromSize is null || fromSize < 1)
        {
            throw Refuse("LOG_MALFORMED", "fromSize must be a positive integer");
        }
        if (toSize is null || toSize < 1)
        {
            throw Refuse("LOG_MALFORMED", "toSize must be a positive integer");
        }
        if (proof is null)
        {
            throw Refuse("LOG_MALFORMED", "proof must be an array of node hashes");
        }
        var nodes = proof.Select((value, index) => HashFromHex(value, $"proof[{index}]")).ToList();

        if (fromSize > toSize)
        {
            throw Refuse("CONSISTENCY_PROOF_INVALID", $"a tree of {toSize} leaves cannot extend a tree of {fromSize}");
        }
        if (fromSize == toSize)
        {
            if (nodes.Count != 0)
            {
                throw Refuse("CONSISTENCY_PROOF_INVALID", "equal sizes take an empty proof");
            }
            if (!fromRoot.AsSpan().SequenceEqual(toRoot))
            {
                throw Refuse("CONSISTENCY_PROOF_INVALID", "two heads at one size disagree — the tree was altered in place");
            }
            return;
        }
        if ((fromSize & (fromSize - 1)) == 0)
        {
            nodes.Insert(0, fromRoot);
        }
        if (nodes.Count == 0)
        {
            throw Refuse("CONSISTENCY_PROOF_INVALID", "the proof is empty");
        }

        var fn = fromSize.Value - 1;
        var sn = toSize.Value - 1;
        while (fn % 2 == 1)
        {
            fn /= 2;
            sn /= 2;
        }
        var fr = nodes[0];
        var sr = nodes[0];
        foreach (var c in nodes.Skip(1))
        {
            if (sn == 0)
            {
                throw Refuse("CONSISTENCY_PROOF_INVALID", "the proof is longer than the new tree is deep");
            }
            if (fn % 2 == 1 || fn == sn)
            {
                fr = Node(c, fr);
                sr = Node(c, sr);
                if (fn % 2 == 0)
                {
                    while (fn != 0 && fn % 2 == 0)
                    {
                        fn /= 2;
                        sn /= 2;
                    }
                }
            }
            else
            {
                sr = Node(sr, c);
            }
            fn /= 2;
            sn /= 2;
        }
        if (sn != 0)
        {
            throw Refuse("CONSISTENCY_PROOF_INVALID", "the proof is shorter than the new tree is deep");
        }
        if (!fr.AsSpan().SequenceEqual(fromRoot))
        {
            throw Refuse("CONSISTENCY_PROOF_INVALID", "the proof does not reproduce the old head");
        }
        if (!sr.AsSpan().SequenceEqual(toRoot))
        {
            throw Refuse("CONSISTENCY_PROOF_INVALID", "the proof does not reproduce the new head");
        }
    }

    // -- Checkpoints ---------------------------------------------------------

    private static readonly string[] CheckpointMembers =
    [
        "assuranceProfile", "checkpointSequence", "issuedAt", "origin", "rootHash", "signer", "treeSize", "version",
    ];
    private static readonly string[] SignerMembers = ["keyId", "provider", "suite"];

    public static JsonObject ValidateCheckpointDocument(JsonNode? node)
    {
        if (node is not JsonObject document)
        {
            throw Refuse("LOG_MALFORMED", "a checkpoint must be an object");
        }
        RequireSupported(Canonical.StringOf(document, "version"));
        var unknown = document.Select(pair => pair.Key).Where(member => !CheckpointMembers.Contains(member)).ToList();
        if (unknown.Count > 0)
        {
            throw Refuse("LOG_MALFORMED", $"a checkpoint carries {string.Join(", ", unknown)}; the member set is closed");
        }
        var missing = CheckpointMembers.Where(member => !document.ContainsKey(member)).ToList();
        if (missing.Count > 0)
        {
            throw Refuse("LOG_MALFORMED", $"a checkpoint is missing {string.Join(", ", missing)}");
        }
        if (!AssuranceProfiles.Contains(Canonical.StringOf(document, "assuranceProfile")))
        {
            throw Refuse("LOG_MALFORMED", $"assuranceProfile must be one of {string.Join(", ", AssuranceProfiles)}");
        }
        SafePositiveInteger(document["checkpointSequence"], "checkpointSequence");
        SafePositiveInteger(document["treeSize"], "treeSize");
        BoundText(document["origin"], "origin");
        var rootHash = Canonical.StringOf(document, "rootHash");
        if (rootHash is null || !RootHashPattern.IsMatch(rootHash))
        {
            throw Refuse("LOG_MALFORMED", "rootHash must be \"sha256:\" then 64 lowercase hex digits");
        }
        var issuedAt = Canonical.StringOf(document, "issuedAt");
        string? issuedAtCanonical = null;
        if (issuedAt is not null)
        {
            try
            {
                issuedAtCanonical = Canonical.CanonicalTimestamp(issuedAt);
            }
            catch (CommitmentException)
            {
                issuedAtCanonical = null;
            }
        }
        if (issuedAtCanonical != issuedAt || issuedAt is null)
        {
            throw Refuse("LOG_MALFORMED", "issuedAt must be a canonical UTC timestamp");
        }
        if (document["signer"] is not JsonObject signer)
        {
            throw Refuse("LOG_MALFORMED", "signer must be an object");
        }
        var signerUnknown = signer.Select(pair => pair.Key).Where(member => !SignerMembers.Contains(member)).ToList();
        if (signerUnknown.Count > 0)
        {
            throw Refuse("LOG_MALFORMED", $"signer carries {string.Join(", ", signerUnknown)}");
        }
        BoundText(signer["keyId"], "signer.keyId");
        BoundText(signer["provider"], "signer.provider");
        var suite = Canonical.StringOf(signer, "suite");
        if (suite is null || !Registry.Suites.ContainsKey(suite))
        {
            throw Refuse("LOG_MALFORMED", $"signer.suite \"{suite}\" is not a suite this contract defines");
        }
        return document;
    }

    public static byte[] CheckpointSigningInput(JsonNode? node)
    {
        var document = ValidateCheckpointDocument(node);
        var header = string.Join(Separator, CheckpointDomainTag, Canonical.StringOf(document, "version"), "");
        return [.. Encoding.UTF8.GetBytes(header), .. CanonicalBytes(document, "a checkpoint")];
    }

    public static void VerifyCheckpoint(JsonNode? checkpointNode, JsonNode? signatureNode, JsonObject jwk)
    {
        var checkpoint = ValidateCheckpointDocument(checkpointNode);
        var suiteId = Canonical.StringOf((JsonObject)checkpoint["signer"]!, "suite")!;
        var suite = Registry.Suites[suiteId];
        var signatureBytes = DecodeSignature(signatureNode, "signature");
        if (signatureBytes.Length != suite.SignatureLength)
        {
            throw Refuse("MALFORMED_SIGNATURE", $"signature is {signatureBytes.Length} bytes");
        }
        try
        {
            using var key = Registry.LoadPublicKey(jwk, suiteId);
        }
        catch (Exception exception) when (exception is CryptographicException or TrustRegistryException or ArgumentException or InvalidOperationException)
        {
            throw Refuse("INVALID_PUBLIC_KEY", "jwk is not a valid public key");
        }
        if (!Registry.SignatureVerifies(CheckpointSigningInput(checkpoint), jwk, suiteId, signatureBytes))
        {
            throw Refuse("CHECKPOINT_SIGNATURE_INVALID", "the signature over this checkpoint does not verify under this key");
        }
    }

    public static JsonObject AcceptCheckpoint(JsonObject? accepted, JsonNode? candidateNode, JsonArray? consistencyProof)
    {
        var candidate = ValidateCheckpointDocument(candidateNode);
        var state = new JsonObject
        {
            ["checkpointSequence"] = Registry.IntegerOf(candidate["checkpointSequence"])!.Value,
            ["origin"] = Canonical.StringOf(candidate, "origin"),
            ["rootHash"] = Canonical.StringOf(candidate, "rootHash"),
            ["treeSize"] = Registry.IntegerOf(candidate["treeSize"])!.Value,
        };
        if (accepted is null)
        {
            return state;
        }
        BoundText(accepted["origin"], "accepted.origin");
        SafePositiveInteger(accepted["checkpointSequence"], "accepted.checkpointSequence");
        SafePositiveInteger(accepted["treeSize"], "accepted.treeSize");
        var acceptedRoot = Canonical.StringOf(accepted, "rootHash");
        if (acceptedRoot is null || !RootHashPattern.IsMatch(acceptedRoot))
        {
            throw Refuse("LOG_MALFORMED", "accepted.rootHash must be \"sha256:\" then 64 lowercase hex digits");
        }

        var candidateOrigin = Canonical.StringOf(candidate, "origin");
        var acceptedOrigin = Canonical.StringOf(accepted, "origin");
        if (candidateOrigin != acceptedOrigin)
        {
            throw Refuse("CHECKPOINT_ORIGIN_MISMATCH", $"this checkpoint is from log \"{candidateOrigin}\", not \"{acceptedOrigin}\"");
        }
        var candidateSequence = Registry.IntegerOf(candidate["checkpointSequence"])!.Value;
        var acceptedSequence = Registry.IntegerOf(accepted["checkpointSequence"])!.Value;
        var candidateSize = Registry.IntegerOf(candidate["treeSize"])!.Value;
        var acceptedSize = Registry.IntegerOf(accepted["treeSize"])!.Value;
        var candidateRoot = Canonical.StringOf(candidate, "rootHash")!;
        if (candidateSequence < acceptedSequence)
        {
            throw Refuse("CHECKPOINT_STALE", $"checkpoint {candidateSequence} is older than the already accepted {acceptedSequence}");
        }
        if (candidateSequence == acceptedSequence)
        {
            if (candidateSize == acceptedSize && candidateRoot == acceptedRoot)
            {
                return state;
            }
            throw Refuse("CHECKPOINT_FORK", $"two checkpoints numbered {candidateSequence} disagree — the log presented two histories");
        }
        if (candidateSize < acceptedSize)
        {
            throw Refuse("CHECKPOINT_ROLLBACK", $"checkpoint {candidateSequence} covers {candidateSize} leaves but {acceptedSequence} already covered {acceptedSize}");
        }
        if (candidateSize == acceptedSize)
        {
            if (candidateRoot != acceptedRoot)
            {
                throw Refuse("CHECKPOINT_FORK", "a later checkpoint at the same tree size names a different head");
            }
            return state;
        }
        VerifyConsistency(
            acceptedSize,
            acceptedRoot["sha256:".Length..],
            candidateSize,
            candidateRoot["sha256:".Length..],
            consistencyProof ?? []);
        return state;
    }
}

File: conformance/python/pruvz_verifier/evidence_log.py
"""Append-only evidence log, format version 1 — independent Python implementation.

Implemented from docs/EVIDENCE-LOG.md (PRUVZ-97): RFC 6962 / RFC 9162 tree
composition over domain-separated leaves, signed checkpoints, and the
stateful acceptance rules that turn fork, rollback and stale presentation
into refusals.
"""

from __future__ import annotations

import hashlib
import re
from typing import Any, Optional

from .canonical import canonical_timestamp, canonicalize, CommitmentError
from .registry import SUITES, TrustRegistryError, decode_base64url, load_public_key, signature_verifies

EVIDENCE_LOG_FORMAT_VERSION = "1"
LEAF_ENVELOPE_VERSION = "1"
LEAF_DOMAIN_TAG = "pruvz.ai/evidence-log-leaf"
CHECKPOINT_DOMAIN_TAG = "pruvz.ai/log-checkpoint"
_SEPARATOR = "\u0000"
LEAF_HASH_PREFIX = b"\x00"
NODE_HASH_PREFIX = b"\x01"
ASSURANCE_PROFILES = ("PRE_CUSTOMER_DEFAULT", "CUSTOMER_PRODUCTION")
_MAX_BOUND_TEXT_LENGTH = 512
_MAX_SAFE_INTEGER = 9007199254740991
_HEX = re.compile(r"^[0-9a-f]{64}$")
_ROOT_HASH = re.compile(r"^sha256:[0-9a-f]{64}$")
_PRINTABLE_ASCII = re.compile(r"^[\x20-\x7e]+$")


class EvidenceLogError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


def _fail(code: str, message: str) -> None:
    raise EvidenceLogError(code, message)


def _is_object(value: Any) -> bool:
    return isinstance(value, dict)


def _canonical_bytes(document: Any, context: str) -> bytes:
    try:
        return canonicalize(document)
    except CommitmentError as error:
        _fail("LOG_MALFORMED", f"{context} cannot be canonicalized: {error}")


def _sha256(*parts: bytes) -> bytes:
    digest = hashlib.sha256()
    for part in parts:
        digest.update(part)
    return digest.digest()


def _hash_from_hex(value: Any, field: str) -> bytes:
    if not isinstance(value, str) or _HEX.match(value) is None:
        _fail("LOG_MALFORMED", f"{field} must be 64 lowercase hex digits of SHA-256")
    return bytes.fromhex(value)


def _decode_signature(value: Any, field: str) -> bytes:
    try:
        return decode_base64url(value, field, "MALFORMED_SIGNATURE")
    except TrustRegistryError as error:
        _fail("MALFORMED_SIGNATURE", str(error))


def _bound_text(value: Any, field: str) -> str:
    if not isinstance(value, str) or value == "":
        _fail("LOG_MALFORMED", f"{field} is required")
    if len(value) > _MAX_BOUND_TEXT_LENGTH:
        _fail("LOG_MALFORMED", f"{field} must not exceed {_MAX_BOUND_TEXT_LENGTH} characters")
    if _PRINTABLE_ASCII.match(value) is None:
        _fail("LOG_MALFORMED", f"{field} must be printable ASCII")
    return value


def _safe_positive_integer(value: Any, field: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value < 1 or value > _MAX_SAFE_INTEGER:
        _fail("LOG_MALFORMED", f"{field} must be an integer between 1 and {_MAX_SAFE_INTEGER}")
    return value


def require_supported(format_version: Any) -> None:
    if format_version != EVIDENCE_LOG_FORMAT_VERSION:
        shown = "(none)" if format_version is None else f'"{format_version}"'
        _fail("UNKNOWN_LOG_VERSION", f"Unknown evidence-log version {shown}.")


# ── Leaf encoding ───────────────────────────────────────────────────────────

def leaf_input(seal: Any) -> bytes:
    if not _is_object(seal) or not _is_object(seal.get("envelope")):
        _fail("LOG_MALFORMED", "a leaf must be { envelope, signature }")
    if seal["envelope"].get("version") != LEAF_ENVELOPE_VERSION:
        _fail("LOG_MALFORMED", f"a leaf carries an envelope of version {LEAF_ENVELOPE_VERSION}")
    signature = _decode_signature(seal.get("signature"), "seal.signature")
    suite_id = (seal["envelope"].get("signer") or {}).get("suite") if _is_object(seal["envelope"].get("signer")) else None
    suite = SUITES.get(suite_id) if isinstance(suite_id, str) else None
    if suite is not None and len(signature) != suite["signatureLength"]:
        _fail("MALFORMED_SIGNATURE", f"seal.signature is {len(signature)} bytes")
    unknown = [member for member in seal.keys() if member not in ("envelope", "signature")]
    if unknown:
        _fail("LOG_MALFORMED", f"a leaf holds exactly envelope and signature; got {', '.join(unknown)}")
    header = _SEPARATOR.join([LEAF_DOMAIN_TAG, EVIDENCE_LOG_FORMAT_VERSION, ""])
    return header.encode("utf-8") + _canonical_bytes(seal, "a leaf")


def leaf_hash_of(data: bytes) -> bytes:
    return _sha256(LEAF_HASH_PREFIX, data)


def seal_leaf_hash(seal: Any) -> str:
    return leaf_hash_of(leaf_input(seal)).hex()


# ── Tree composition (RFC 6962 §2.1) ────────────────────────────────────────

def empty_tree_hash() -> str:
    return _sha256().hex()


def _node(left: bytes, right: bytes) -> bytes:
    return _sha256(NODE_HASH_PREFIX, left, right)


def _split(n: int) -> int:
    k = 1
    while k * 2 < n:
        k *= 2
    return k


def _leaf_buffers(leaf_hashes: Any, field: str) -> list:
    if not isinstance(leaf_hashes, list):
        _fail("LOG_MALFORMED", f"{field} must be an array of leaf hashes")
    return [_hash_from_hex(value, f"{field}[{index}]") for index, value in enumerate(leaf_hashes)]


def _head(hashes: list, lo: int, hi: int) -> bytes:
    n = hi - lo
    if n == 1:
        return hashes[lo]
    k = _split(n)
    return _node(_head(hashes, lo, lo + k), _head(hashes, lo + k, hi))


def tree_head(leaf_hashes: Any) -> str:
    hashes = _leaf_buffers(leaf_hashes, "leafHashes")
    if len(hashes) == 0:
        return empty_tree_hash()
    return _head(hashes, 0, len(hashes)).hex()


def verify_inclusion(leaf_hash: Any, leaf_index: Any, tree_size: Any, path: Any, root_hash: Any) -> None:
    leaf = _hash_from_hex(leaf_hash, "leafHash")
    root = _hash_from_hex(root_hash, "rootHash")
    if not isinstance(leaf_index, int) or isinstance(leaf_index, bool) or leaf_index < 0:
        _fail("LOG_MALFORMED", "leafIndex must be a non-negative integer")
    if not isinstance(tree_size, int) or isinstance(tree_size, bool) or tree_size < 1:
        _fail("LOG_MALFORMED", "treeSize must be a positive integer")
    if not isinstance(path, list):
        _fail("LOG_MALFORMED", "path must be an array of node hashes")
    nodes = [_hash_from_hex(value, f"path[{index}]") for index, value in enumerate(path)]

    if leaf_index >= tree_size:
        _fail("INCLUSION_PROOF_INVALID", f"leaf index {leaf_index} does not exist in a tree of {tree_size} leaves")

    fn, sn, r = leaf_index, tree_size - 1, leaf
    for p in nodes:
        if sn == 0:
            _fail("INCLUSION_PROOF_INVALID", "the path is longer than the tree is deep")
        if fn % 2 == 1 or fn == sn:
            r = _node(p, r)
            if fn % 2 == 0:
                while fn != 0 and fn % 2 == 0:
                    fn //= 2
                    sn //= 2
        else:
            r = _node(r, p)
        fn //= 2
        sn //= 2
    if sn != 0:
        _fail("INCLUSION_PROOF_INVALID", "the path is shorter than the tree is deep")
    if r != root:
        _fail("INCLUSION_PROOF_INVALID", "the path does not lead this leaf to this root")


def verify_consistency(from_size: Any, from_root_hash: Any, to_size: Any, to_root_hash: Any, proof: Any) -> None:
    from_root = _hash_from_hex(from_root_hash, "fromRootHash")
    to_root = _hash_from_hex(to_root_hash, "toRootHash")
    if not isinstance(from_size, int) or isinstance(from_size, bool) or from_size < 1:
        _fail("LOG_MALFORMED", "fromSize must be a positive integer")
    if not isinstance(to_size, int) or isinstance(to_size, bool) or to_size < 1:
        _fail("LOG_MALFORMED", "toSize must be a positive integer")
    if not isinstance(proof, list):
        _fail("LOG_MALFORMED", "proof must be an array of node hashes")
    nodes = [_hash_from_hex(value, f"proof[{index}]") for index, value in enumerate(proof)]

    if from_size > to_size:
        _fail("CONSISTENCY_PROOF_INVALID", f"a tree of {to_size} leaves cannot extend a tree of {from_size}")
    if from_size == to_size:
        if len(nodes) != 0:
            _fail("CONSISTENCY_PROOF_INVALID", "equal sizes take an empty proof")
        if from_root != to_root:
            _fail("CONSISTENCY_PROOF_INVALID", "two heads at one size disagree — the tree was altered in place")
        return

    if from_size & (from_size - 1) == 0:
        nodes = [from_root] + nodes
    if len(nodes) == 0:
        _fail("CONSISTENCY_PROOF_INVALID", "the proof is empty")

    fn, sn = from_size - 1, to_size - 1
    while fn % 2 == 1:
        fn //= 2
        sn //= 2

    fr = sr = nodes[0]
    for c in nodes[1:]:
        if sn == 0:
            _fail("CONSISTENCY_PROOF_INVALID", "the proof is longer than the new tree is deep")
        if fn % 2 == 1 or fn == sn:
            fr = _node(c, fr)
            sr = _node(c, sr)
            if fn % 2 == 0:
                while fn != 0 and fn % 2 == 0:
                    fn //= 2
                    sn //= 2
        else:
            sr = _node(sr, c)
        fn //= 2
        sn //= 2

    if sn != 0:
        _fail("CONSISTENCY_PROOF_INVALID", "the proof is shorter than the new tree is deep")
    if fr != from_root:
        _fail("CONSISTENCY_PROOF_INVALID", "the proof does not reproduce the old head")
    if sr != to_root:
        _fail("CONSISTENCY_PROOF_INVALID", "the proof does not reproduce the new head")


# ── Checkpoints ─────────────────────────────────────────────────────────────

_CHECKPOINT_MEMBERS = (
    "assuranceProfile",
    "checkpointSequence",
    "issuedAt",
    "origin",
    "rootHash",
    "signer",
    "treeSize",
    "version",
)
_SIGNER_MEMBERS = ("keyId", "provider", "suite")


def validate_checkpoint_document(document: Any) -> dict:
    if not _is_object(document):
        _fail("LOG_MALFORMED", "a checkpoint must be an object")
    require_supported(document.get("version"))
    unknown = [member for member in document.keys() if member not in _CHECKPOINT_MEMBERS]
    if unknown:
        _fail("LOG_MALFORMED", f"a checkpoint carries {', '.join(unknown)}; the member set is closed")
    missing = [member for member in _CHECKPOINT_MEMBERS if member not in document]
    if missing:
        _fail("LOG_MALFORMED", f"a checkpoint is missing {', '.join(missing)}")
    if document["assuranceProfile"] not in ASSURANCE_PROFILES:
        _fail("LOG_MALFORMED", f"assuranceProfile must be one of {', '.join(ASSURANCE_PROFILES)}")
    _safe_positive_integer(document["checkpointSequence"], "checkpointSequence")
    _safe_positive_integer(document["treeSize"], "treeSize")
    _bound_text(document["origin"], "origin")
    if not isinstance(document["rootHash"], str) or _ROOT_HASH.match(document["rootHash"]) is None:
        _fail("LOG_MALFORMED", 'rootHash must be "sha256:" then 64 lowercase hex digits')
    issued_at_canonical = None
    if isinstance(document["issuedAt"], str):
        try:
            issued_at_canonical = canonical_timestamp(document["issuedAt"])
        except CommitmentError:
            issued_at_canonical = None
    if issued_at_canonical != document["issuedAt"]:
        _fail("LOG_MALFORMED", "issuedAt must be a canonical UTC timestamp")
    signer = document["signer"]
    if not _is_object(signer):
        _fail("LOG_MALFORMED", "signer must be an object")
    signer_unknown = [member for member in signer.keys() if member not in _SIGNER_MEMBERS]
    if signer_unknown:
        _fail("LOG_MALFORMED", f"signer carries {', '.join(signer_unknown)}")
    _bound_text(signer.get("keyId"), "signer.keyId")
    _bound_text(signer.get("provider"), "signer.provider")
    if not isinstance(signer.get("suite"), str) or signer["suite"] not in SUITES:
        _fail("LOG_MALFORMED", f'signer.suite "{signer.get("suite")}" is not a suite this contract defines')
    return document


def checkpoint_signing_input(document: Any) -> bytes:
    validate_checkpoint_document(document)
    header = _SEPARATOR.join([CHECKPOINT_DOMAIN_TAG, document["version"], ""])
    return header.encode("utf-8") + _canonical_bytes(document, "a checkpoint")


def verify_checkpoint(checkpoint: Any, signature: Any, jwk: Any) -> None:
    validate_checkpoint_document(checkpoint)
    suite = SUITES[checkpoint["signer"]["suite"]]
    signature_bytes = _decode_signature(signature, "signature")
    if len(signature_bytes) != suite["signatureLength"]:
        _fail("MALFORMED_SIGNATURE", f"signature is {len(signature_bytes)} bytes")
    if not _is_object(jwk):
        _fail("INVALID_PUBLIC_KEY", "jwk must be a public JSON Web Key object")
    try:
        load_public_key(jwk, checkpoint["signer"]["suite"])
    except TrustRegistryError:
        _fail("INVALID_PUBLIC_KEY", "jwk is not a valid public key")
    if not signature_verifies(checkpoint_signing_input(checkpoint), jwk, checkpoint["signer"]["suite"], signature_bytes):
        _fail("CHECKPOINT_SIGNATURE_INVALID", "the signature over this checkpoint does not verify under this key")


def accept_checkpoint(accepted: Any, candidate: Any, consistency_proof: Any = None) -> dict:
    validate_checkpoint_document(candidate)
    state = {
        "checkpointSequence": candidate["checkpointSequence"],
        "origin": candidate["origin"],
        "rootHash": candidate["rootHash"],
        "treeSize": candidate["treeSize"],
    }
    if accepted is None:
        return state
    if not _is_object(accepted):
        _fail("LOG_MALFORMED", "accepted must be the state a previous acceptance returned")
    _bound_text(accepted.get("origin"), "accepted.origin")
    _safe_positive_integer(accepted.get("checkpointSequence"), "accepted.checkpointSequence")
    _safe_positive_integer(accepted.get("treeSize"), "accepted.treeSize")
    if not isinstance(accepted.get("rootHash"), str) or _ROOT_HASH.match(accepted["rootHash"]) is None:
        _fail("LOG_MALFORMED", 'accepted.rootHash must be "sha256:" then 64 lowercase hex digits')

    if candidate["origin"] != accepted["origin"]:
        _fail("CHECKPOINT_ORIGIN_MISMATCH", f'this checkpoint is from log "{candidate["origin"]}", not "{accepted["origin"]}"')
    if candidate["checkpointSequence"] < accepted["checkpointSequence"]:
        _fail("CHECKPOINT_STALE", f"checkpoint {candidate['checkpointSequence']} is older than the already accepted {accepted['checkpointSequence']}")
    if candidate["checkpointSequence"] == accepted["checkpointSequence"]:
        if candidate["treeSize"] == accepted["treeSize"] and candidate["rootHash"] == accepted["rootHash"]:
            return dict(state)
        _fail("CHECKPOINT_FORK", f"two checkpoints numbered {candidate['checkpointSequence']} disagree — the log presented two histories")
    if candidate["treeSize"] < accepted["treeSize"]:
        _fail("CHECKPOINT_ROLLBACK", f"checkpoint {candidate['checkpointSequence']} covers {candidate['treeSize']} leaves but {accepted['checkpointSequence']} already covered {accepted['treeSize']}")
    if candidate["treeSize"] == accepted["treeSize"]:
        if candidate["rootHash"] != accepted["rootHash"]:
            _fail("CHECKPOINT_FORK", "a later checkpoint at the same tree size names a different head")
        return state
    verify_consistency(
        accepted["treeSize"],
        accepted["rootHash"][len("sha256:"):],
        candidate["treeSize"],
        candidate["rootHash"][len("sha256:"):],
        consistency_proof if consistency_proof is not None else [],
    )
    return state
